paper/scripts/run_exp_by_examples.py

```diff
@@ -0,0 +1,294 @@
+        load_dataset = getattr(datasets, 'get_'+checkpoint['dataset_name'])
+        train_loader, _, test_loader, mem_loader = load_dataset(dataset_dir, batch_size_train=256,
+                batch_size_test=256, batch_size_memory=mem_examples, size_train=train_examples,
+                seed=run)
+
+        model.eval()
+
+        # Final CNN weights for computing twin contributions
+        if modality == 'std':
+            weights = model.linear.weight
+        else:
+            weights = model.mw.classifier.fc2.weight
+
+
+        train_conts = list()
+        train_preds = list()
+        train_encoded = list()
+        train_inputs = list()
+
+
+        if metric == 'input':
+            loss_criterion = torch.nn.L1Loss()
+        else:
+            loss_criterion = torch.nn.CrossEntropyLoss()
+
+        # GLOBAL
+        with torch.no_grad():
+            for i, data in enumerate(train_loader):
+                imgs, _ = data
+                imgs = imgs.to(device)
+                if modality == 'std':
+                    logits, encoded_imgs= model(imgs)
+                else:
+                    memory, _ = next(iter(mem_loader))
+                    memory = memory.to(device)
+                    logits, encoded_imgs= model(imgs,memory)
+                pred = torch.argmax(logits, axis=1)
+
+                x_cont = weights[pred] * encoded_imgs
+
+                train_conts.extend(x_cont)
+                train_preds.extend(pred)
+                train_encoded.extend(encoded_imgs)
+                train_inputs.extend(imgs)
+        train_inputs = torch.stack(train_inputs)
+        train_encoded = torch.stack(train_encoded)
+        train_conts = torch.stack(train_conts)
+        train_preds = torch.stack(train_preds)
+
+
+        # test knn
+        chpt1_differences = []
+        mem_differences = []
+        dknn1_differences = []
+        random_difference = []
+        tot = 0
+        with torch.no_grad():
+            for i, data in enumerate(test_loader):
+                # reset the data to train k-NN
+                train_conts = list()
+                train_preds = list()
+                imgs, _ = data
+                imgs = imgs.to(device)
+                tot += len(imgs)
+                if modality == 'std':
+                    # launch an exception if the model is not a std model
+                    raise Exception('The model is not a std model')
+                else:
+                    memory, _ = next(iter(mem_loader))
+                    memory = memory.to(device)
+                    logits, encoded_imgs, memory_weights = model(imgs,memory,return_weights=True)
+                pred = torch.argmax(logits, axis=1)
+
+                # fit knn using only memory samples
+                aux_memory, _ = next(iter(mem_loader))
+                aux_memory = aux_memory.to(device)
+                exp_logits, exp_x = model(memory,aux_memory)
+                exp_pred =  torch.argmax(exp_logits, axis=1)
+
+
```

Comment:
too many emtpy lines

Reply:
fixed in [25aa7c8](https://github.com/KRLGroup/memory-wrap/pull/3/commits/25aa7c89c4a501dde33fa10bf3d0f5f29da27ca2)